### ISSUE TITLE
Modules and concatenation strategy

### ISSUE THREAD

**Issue by potomak** [craftyjs/Crafty#557]
I'd like to remove intro.js, outro.js, and to _fix_ core.js, see #556.

/cc @louisstow @starwed

**Comment by starwed** [comment]
Ah, I probably wouldn't use the term "loading" to describe this.  We concatenate all the files together to produce crafty.js, and a release is of that single file.    Someone just writing a game would never see intro.js.

It would be great if each file could also stand alone (since then we could run e.g. jsvalidate against them each separately) but only if it's strictly better than what we do right now.

**Comment by sbyrnes321** [comment]
I agree with @starwed -- I don't think cancatenating is a bad solution. We can just remember that those jshints should be ignored...

**Comment by potomak** [comment]
I partially agree, I think we're also speaking about loading as right now if you want to develop or test the library you must do it against the compiled version.

Using modules we can test them separately and use them separately during development.

Moreover right now compiling is order dependent, I think it's not a reliable solution.

Note: I think that

> We can just remember that those jshints should be ignored...

it's not the best solution.

**Comment by sbyrnes321** [comment]
jquery has an intro and outro -- https://github.com/jquery/jquery/blob/master/src/outro.js -- it can't be _such_ a bad idea.

If you don't like my idea of ignoring those jshint warnings, here's another idea: The grunt jshint task can be edited to avoid jshint'ing the very simple files `intro.js` and `outro.js`, and to ignore the specific error "E006: Unexpected early end of program." in the specific file `core.js`. Then the jshint task can run perfectly with no warnings displayed.

I disagree with your comment "compiling is order dependent, I think it's not a reliable solution." Concatenation has to be done in the order core, intro, (everything else), outro. What is "unreliable"? It's true that someone could change the order of the list in package.json. But then they would immediately notice that this change makes crafty stop working! So they would undo that change, and put the order back to the way it was. If you're really concerned about this, you can add a comment in package.json:

```
// The order of the files in this list needs to be core.js, intro.js, (everything else), outro.js,
// because the files are concatenated in this order to create crafty.js
```

Then you can be 100% confident that the files will always be concatenated in the right order. Very "reliable", right?

Again, I don't see any problem with the status quo. If you have another idea, can you describe more specifically how it would work?

**Comment by potomak** [comment]
My idea is to rewrite modules using the ES6 draft specification than using [ES6 Module Transpiler](http://square.github.io/es6-module-transpiler/) compile them into an existing library-based module systems such as AMD and finally build them using a tool such as [grunt-contrib-requirejs](https://github.com/gruntjs/grunt-contrib-requirejs)

**Comment by kenany** [comment]
I agree with what @potomak is proposing. Crafty's source code would be easier to read (and more scalable) if each of the files in `src/` clearly defined its dependencies to the other files.

For example, many of the source files contain something like this:

``` javascript
Crafty.c("Collision", {
```

Here's the problem: it's not immediately obvious to me where this `Crafty` "global" came from. If each file defined its dependencies, it would make much more sense:

``` javascript
var Crafty = require('./core.js');

Crafty.c("Collision", {
```

Whether AMD/CommonJS/ES6 is used is up to the preferences of you guys.

**Comment by sbyrnes321** [comment]
@potomak -- I'm not familiar with those things, so I have a basic question.

In the current build process, we end up with a single file called `crafty.js`, which does not rely on any other javascript library, and which can be loaded directly or via commonjs, requirejs, etc., with no compatibility problems.

Will that still be true after your proposed change?

**Comment by potomak** [comment]
@sbyrnes321 yes.

That's the exact purpose of r.js (the [RequireJS Optimizer](http://requirejs.org/docs/optimization.html)). You can use it through `grunt-contrib-requirejs` task and it concatenates and optimizes the code.
The result of this task will be a `crafty.js` file that you can include to use the Crafty engine as it is now.

See also: http://requirejs.org/docs/faq-optimization.html#wrap

**Comment by sbyrnes321** [comment]
@potomak -- OK, thanks. Your proposal sounds like it will work fine, as far as I can tell.

**Comment by kevinsimper** [comment]
I think the CommonJS approuch would be better and simpler to understand, since you would not to create AMD modules.

Then with browserify you would create a bundled file.

**Comment by potomak** [comment]
I agree about this the simplicity of CommonJS, but I never used browserify.

**Comment by kevinsimper** [comment]
Fixed in #568
